### ISSUE TITLE
[InstCombine] `!(A & INT_MIN) && A & INT_MAX` is not folded to signed comparison with zero

### ISSUE THREAD

**Issue by artem** [llvm/llvm-project#69123]
https://godbolt.org/z/rxPnbEnoT

```c
void fn();

void test(unsigned *d) {
    int is_negative = *d & 0x80000000;
    
    if (!is_negative && *d & 0x7fffffff) {
        fn();
    }
}
```

GCC recognizes this pattern and folds it to comparison 

```asm
test:
        mov     eax, DWORD PTR [rdi]
        test    eax, eax
        jg      .L4
        ret
.L4:
        xor     eax, eax
        jmp     fn
```

However, LLVM 18 main fails to optimize it:

```asm
test:
        mov     eax, dword ptr [rdi]
        test    eax, eax
        js      .LBB0_2
        and     eax, 2147483647
        je      .LBB0_2
        xor     eax, eax
        jmp     fn@PLT
.LBB0_2:
        ret
```

**Comment by dtcxzyw** [comment]
Alive2: https://alive2.llvm.org/ce/z/JDWun8

**Comment by artem** [comment]
Hmm, this seems to be fixed as of the current main branch

**Comment by artem** [comment]
Fix bisected to #73662
